lib/brexit_checker/criteria/filter.rb

```diff
@@ -0,0 +1,32 @@
+class BrexitChecker::Criteria::Filter
+  def call(selected_criteria)
+    loop do
+      old_size = selected_criteria.size
+      selected_criteria &&= possible_criteria(selected_criteria)
+      break if selected_criteria.size == old_size
+    end
+
+    selected_criteria
+  end
+
+private
+
+  def possible_criteria(selected_criteria)
+    BrexitChecker::Question.load_all
+      .select { |question| question.show?(selected_criteria) }
+      .flat_map(&:options)
+      .flat_map { |option| possible_values(option, selected_criteria) }
+      .compact
+  end
+
+  def possible_values(option, selected_criteria)
+    return [] unless option.value.blank? ||
+      selected_criteria.include?(option.value)
```

Comment:
Indeed. I wasn't particularly happy with the name. I think I've discussed this already in replying to a previous comment. Briefly: I did try to use `all_values` to help the code here, but it turned into a mess because of the tree-based filtering aspect.